server/backend/cafeteriaService.go

```diff
@@ -0,0 +1,687 @@
+	var result []int
+	err := s.db.Model(&cafeteria_rating_models.DishToDishNameTag{}).Where("dishID = ? ", dishID).Select("nameTagID").Scan(&result).Error
+	if err != nil {
+		log.Error(err)
+	} else {
+		for _, tagID := range result {
+			s.db.Model(&cafeteria_rating_models.DishNameTag{}).Create(&cafeteria_rating_models.DishNameTag{
+				CorrespondingRating: rating.DishRating,
+				Points:              rating.Points,
+				TagNameID:           tagID,
+			})
+		}
+	}
+}
+
+/*
+Checks parameters of the new rating for all cafeteria and dish ratings.
+Additionally, queries the cafeteria ID, since it checks whether the cafeteria actually exists.
+*/
+func inputSanitization(rating int32, image []byte, comment string, cafeteriaName string, s *CampusServer) (int32, error) {
+	if rating > 5 || rating < 0 {
+		return -1, status.Errorf(codes.InvalidArgument, "Rating must be a positive number not larger than 10. Rating has not been saved.")
+	}
+
+	if len(image) > 131100 {
+		return -1, status.Errorf(codes.InvalidArgument, "Image must not be larger than 1MB. Rating has not been saved.")
+	}
+
+	if len(comment) > 256 {
+		return -1, status.Errorf(codes.InvalidArgument, "Ratings can only contain up to 256 characters, this is too long. Rating has not been saved.")
+	}
+
+	if strings.Contains(comment, "@") {
+		return -1, status.Errorf(codes.InvalidArgument, "Comments must not contain @ symbols in order to prevent misuse. Rating has not been saved.")
+	}
+
+	var result *cafeteria_rating_models.Cafeteria
+	testCanteen := s.db.Model(&cafeteria_rating_models.Cafeteria{}).
+		Where("name LIKE ?", cafeteriaName).
+		First(&result)
+	if testCanteen.Error != nil || testCanteen.RowsAffected == 0 {
+		return -1, status.Errorf(codes.InvalidArgument, "Cafeteria does not exist. Rating has not been saved.")
+	}
+
+	return result.Cafeteria, nil
+}
+
+/*
+Checks whether the rating-tag name is a valid option and if so,
+it will be saved with a reference to the rating
+*/
+func storeRatingTags(s *CampusServer, parentRatingID int32, tags []*pb.TagRating, tagType int) (*emptypb.Empty, error) {
+	var errorOccured = ""
+	if len(tags) > 0 {
+		usedTagIds := make(map[int]int)
+		insertModel := getModelStoreTag(tagType, s.db)
+		for _, tag := range tags {
+			var currentTag int
+
+			exists := getModelStoreTagOption(tagType, s.db).
+				Where("EN LIKE ? OR DE LIKE ?", tag.Tag, tag.Tag).
+				Select("id").
+				First(&currentTag)
+
+			if exists.Error != nil || exists.RowsAffected == 0 {
+				log.Println("Tag with tagname ", tag.Tag, "does not exist")
+				errorOccured = errorOccured + ", " + tag.Tag
+			} else {
+				if usedTagIds[currentTag] == 0 {
+					insertModel.
+						Create(&cafeteria_rating_models.DishRatingTag{
+							CorrespondingRating: parentRatingID,
+							Points:              int32(tag.Points),
+							TagID:               currentTag})
+					usedTagIds[currentTag] = 1
+				} else {
+					log.Println("Each Rating tag must be used at most once in a rating.")
```

Comment:
This message looks like it's intended for the user but is printed to the backend logs.

Reply:
This will now be returned as a warning inside the error part of the grpc reply